powernap/man/powernap.8
.TH powernap 8 "2 Jul 2009" powernap "powernap"
.SH NAME
powernap - have the system "take a power nap"; this can be a user defined ACTION_METHOD in \fI/etc/powernap/config\fP among [best-effort, suspend, hibernate, poweroff, powersave]


.SH DESCRIPTION
powernap will run when \fBpowernapd\fP(8) has determined that an action needs to be taken according to its configured parameters. The action to be taken, among [best-effort, suspend, hibernate, poweroff], is determined in \fI/etc/powernap/config\fP. By default, the action will be \fBbest-effort\fP.

When the ACTION_METHOD is \fBbest-effort\fP, and if the file \fI/etc/powernap/action\fP is executable, then the powernap binary will run it when called. Otherwise, powernap will run one of:
  * pm-suspend
  * pm-hibernate
  * poweroff
depending on your hardware's capabilities, as determined by \fBpm-is-supported\fP(1).

You may do one of:
  1) Write your own custom at \fI/etc/powernap/action\fP and make it executable
  2) Replace \fI/etc/powernap/action\fP with an executable script or binary
  3) Symlink \fI/etc/powernap/action\fP to some other executable script or binary

When ACTION_METHOD is \fBpowersave\fP, then powernap binary will execute \fBpm-suspend\fP.

When ACTION_METHOD is \fBpowersave\fP, then powernap binary will execute \fBpm-hibernate\fP.

When ACTION_METHOD is \fBpowersave\fP, then powernap binary will execute \fBpoweroff\fP.

When ACTION_METHOD is \fBpowersave\fP, then powernap binary will execute \fBpm-powersave\fP and pass an argument (true) to execute all the available scripts at \fI/usr/lib/pm-utils/power.d/\fP and \fI/etc/pm/power.d/\fP. The latter will contain scripts shipped with powernap.

.TP
\fIhttp://launchpad.net/powernap\fP
.PD

.SH FILES
\fI/etc/powernap/action\fP, \fI/proc/acpi/sleep\fP

.SH SEE ALSO
\fBpowernapd\fP(8), \fBpm-is-supported\fP(1), \fBethtool\fP(8)

.SH AUTHOR
This manpage and the utility was written by Dustin Kirkland <kirkland@canonical.com> for Ubuntu systems (but may be used by others).  Permission is granted to copy, distribute and/or modify this document under the terms of the GNU General Public License, Version 3 published by the Free Software Foundation.

On Debian systems, the complete text of the GNU General Public License can be found in /usr/share/common-licenses/GPL.


powernap/ChangeLog
powernap (2.18) released; urgency=low

  [ Frank Steinberg ]
  * LoadMonitor: handle float values correctly (LP: #1070696)
  * TCPMonitor: handle IPv6 connections correctly (LP: #1070695)

  [ Andres Rodriguez ]
  * debian/powernap.default: Add so that powernap does not start by default.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Mon, 04 Jun 2012 12:59:15 -0400

powernap (2.17-0ubuntu1) precise; urgency=low

  [ Bryce Harrington ]
  * actions/{service,kernel_module}: Add scripts to handle services and
    kernel modules on powersave mode.

  [ Andres Rodriguez ]
  * Re-organize source to differ between pm-utils actions and powernap
    actions.
    - debian/powernap-common.install: Update.
    - powernap.spec: Update.
  * Disable SERVICES and KERN_MODULES on PowerSave mode (LP: #887684):
    - powernap/powernap.py: Load from config.
    - debian/powernap-common.install: Install new scripts.
    - Add take powersave action and its recover equivalent.
  * actions/pm/01cpu_online: Disable by default.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Thu, 01 Dec 2011 12:43:51 -0500

powernap (2.16-0ubuntu1) precise; urgency=low

  [ Andres Rodriguez ]
  * debian/control: Move python-scapy Depends to Suggests.
  * Do not use awk to parse output. (LP: #879591)
    - bin/powerwake
    - powerwake/powerwake.py

  [ Eric Sandeen ]
  * powernap.spec: cleaned up a bit for better building/installation on RH
    systems

  [ Remi Druilhe ]
  * Dynamically reload configuration file (LP: #887680)

 -- Andres Rodriguez <andreserl@ubuntu.com>  Tue, 29 Nov 2011 14:42:41 -0500

powernap (2.15-0ubuntu1) oneiric; urgency=low

  * actions/eth_speed: Disable by default to not experience connectivity
    loss on certain network configs. (LP: #730210)

 -- Andres Rodriguez <andreserl@ubuntu.com>  Fri, 09 Sep 2011 15:32:30 -0400

powernap (2.14-0ubuntu1) oneiric; urgency=low

  * bin/powerwake:
    - fix the powerake mac caching methodology, which was somewhat broken

 -- Dustin Kirkland <kirkland@ubuntu.com>  Mon, 22 Aug 2011 00:38:17 -0500

powernap (2.13-0ubuntu1) oneiric; urgency=low

  [ Dustin Kirkland ]
  * sbin/powernapd:
    - make sure that powerwake-now works, even if a system has gone into
      powersave mode by other means (erhm, gnome-power-settings, erhm)
  * actions/01cpu_online:
    - drop the nasty sleep 10 hack around gnome-settings-daemon bugginess;
      doesn't seem to be a problem at this point in Oneiric;  If this bug
      shows back up, we need to solve it in a better way
  * actions/video:
    - drop incorrect inline docs
  * config:
    - now that the default is powersave mode, adjust the absent time and
      grace period; it takes fractions of a second to go in and out of
      powersave mode
  * actions/00flag, debian/powernap.upstart, sbin/powernapd: LP: #829618
    - use a flag in shared memory that notes the powersave status of the
      system
    - set/remove that flag in the 00flag script in the pm-utils suite
    - set/remove/query that same flag in powernapd itself
  * debian/control: LP: #637635
    - depend on anacron, to better handle server cronjobs on systems that
      might have been asleep
  * powernap/monitors/InputMonitor.py: LP: #826278
    - test device path existence before opening, fixes nasty crash

 -- Dustin Kirkland <kirkland@ubuntu.com>  Fri, 19 Aug 2011 09:57:07 -0700

powernap (2.12-0ubuntu1) oneiric; urgency=low

  [ Andres Rodriguez ]
  * Add Initial PowerNap (powerwaked) Server Capabilities:
    - powerwake/*: Add module and Initial ARPMonitor (LP: #737479).
    - powerwaked.conf: Add config file.
    - powerwaked: Add daemon.
    - powerwake-monitor: Add tool to add, delete, list, etc.
  * Package powernap-server:
    - debian/control: Add powerwaked, powerwake-common, powernap-server
    - debian/{powerwaked,powerwake-common}.install: Add
    - debian/powerwaked.upstart: Add upstart job.

  [ Dustin Kirkland ]
  * bin/powerwake:
    - only seed the arp cache if the argument is not already a mac address

 -- Andres Rodriguez <andreserl@ubuntu.com>  Tue, 09 Aug 2011 15:28:35 -0400

powernap (2.11-0ubuntu1) oneiric; urgency=low

  * actions/01cpu_online: Delay disabling of CPU's when machine is Desktop
    to allow it to apply template and not crash.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Mon, 20 Jun 2011 18:08:09 -0400

powernap (2.10-0ubuntu1) oneiric; urgency=low

  * Add support for port ranges to TCPMonitor (LP: #760729)
    - powernap/monitors/TCPMonitor.py: Add support.
    - powernap/powernap.py: Ensure that ports are now passed as strings.
    - config: Update to reflect new feature.
  * debian/control:
    - Bump Build-Depends on debhelper.
    - Change python supported version.
    - Update Bzr-VCS to current.
    - Set Dustin as Original maintainer and set myself as maintainer.

 -- Andres Rodriguez <andreserl@ubuntu.com>  Thu, 09 Jun 2011 16:42:59 -0400

powernap (2.9-0ubuntu1) oneiric; urgency=low

  * sbin/powernap, sbin/powernapd:
    - add an indicator to show when powernap is engaged, remove after waking
  * === added directory debian/source, debian/source/format:
    - add this back in (got a fix for release scripts)

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 02 Jun 2011 17:43:23 -0400

powernap (2.8-0ubuntu1) oneiric; urgency=low

  [ Andres Rodriguez ]
  * Switch to dpkg-source 3.0 (quilt) format
  * Switch to dh_python2 instead of dh_pycentral
  * sbin/powernap: ACTION_METHOD is now passed as an argument.
  * Second Stage Action Method (LP: #711521)
    - config: Add config options.
    - powernap/powernap.py: Add loading of config values.
    - sbin/powernapd: Add logic to handle second stage action.

  [ Dustin Kirkland ]
  * config: remove trailing whitespace
  * debian/source/format, === removed directory debian/source:
    - drop these for now, as they're breaking the release scripts

 -- Dustin Kirkland <kirkland@ubuntu.com>  Thu, 02 Jun 2011 16:51:20 -0400

powernap (2.7-0ubuntu1) oneiric; urgency=low


powernap/powernap/powernap.py
#!/usr/bin/python
#
#    powernap.py - handles powernap's config and initializes Monitors.
#
#    Copyright (C) 2010 Canonical Ltd.
#
#    Authors: Andres Rodriguez <andreserl@ubuntu.com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, version 3 of the License.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import configparser, sys, re, os
from .monitors import ProcessMonitor, LoadMonitor, InputMonitor, TCPMonitor, UDPMonitor, IOMonitor, WoLMonitor, ConsoleMonitor, DiskMonitor


class PowerNap:

    def __init__(self):
        self.PKG = "powernap"
        self.CONFIG = "/etc/powernap/config"
        self.ACTION = "/usr/sbin/powernap"
        self.RECOVER_ACTION = "/usr/sbin/pm-powersave false"
        self.ABSENT_SECONDS = sys.maxsize
        self.STAGE2_ABSENT_SECONDS = sys.maxsize
        self.INTERVAL_SECONDS = int(1)
        self.GRACE_SECONDS = int(60)
        self.DEBUG = int(0)
        self.ACTION_METHOD = 0
        self.STAGE2_ACTION_METHOD = 4
        self.MONITORS = []
        self.WARN = False
        self.WATCH_CONFIG = False
        self.KERN_MODULES = {}
        self.SERVICES = {}
        self.stage2_action_enabled = False
        # Load default config file (/etc/powernap/config)
        self.load_config_file()

    def load_config_file(self):
        stage2_section = "%s-stage2" % self.PKG
        cfg = configparser.ConfigParser()
        cfg.read(self.CONFIG)

        try:
            # Load items in DEFAULT section
            defaults = cfg.items(self.PKG)
            for items in defaults:
                self.set_default_values(items[0], items[1])

            stage2 = cfg.items(stage2_section)
            for items in stage2:
                self.set_stage2_values(items[0], items[1])

            # Load items on each monitor
            monitors_config = cfg.sections()
            for monitor in monitors_config:
                if monitor not in [self.PKG, stage2_section]:
                    for items in cfg.items(monitor):
                        self.load_monitors_config(monitor, items)
        except:
            pass

        # Load extra config files (/etc/powernap/config.d/*)
        configd = "/etc/%s/config.d" % self.PKG
        if os.path.exists(configd):
            for config in os.listdir(configd):
                self.load_configd_files("%s/%s" % (configd, config))

    def load_configd_files(self, config_file):
        cfg = configparser.ConfigParser()
        cfg.read(config_file)

        try:
            monitors_config = cfg.sections()
            for monitor in monitors_config:
                for items in cfg.items(monitor):
                    for i in range(len(self.MONITORS)):
                        if self.MONITORS[i]['monitor'] == monitor and self.MONITORS[i]['name'] == items[0]:
                            self.MONITORS.pop(i)
                            break
                    self.load_monitors_config(monitor, items)
        except:
            pass

    def set_default_values(self, var, value):
        if var == "absent_seconds":
            self.ABSENT_SECONDS = eval(value)
        if var == "interval_seconds":
            self.INTERVAL_SECONDS = eval(value)
        if var == "grace_seconds":
            self.GRACE_SECONDS = eval(value)
        if var == "debug":
            self.DEBUG = eval(value)
        if var == "action":
            self.ACTION = eval(value)
        if var == "action_method":
            self.ACTION_METHOD = eval(value)
        if var == "warn":
            if value == "y" or value == "yes":
                self.WARN = True
            else:
                self.WARN = False
        if var == "watch_config":
            if value == "y" or value == "yes":
                self.WATCH_CONFIG = True
            else:
                self.WATCH_CONFIG = False
        if var == "kern_modules":
            self.KERN_MODULES = value.split()
        if var == "services":
            self.SERVICES = value.split()

    def set_stage2_values(self, var, value):
        if var == "stage2_action_method":
            self.STAGE2_ACTION_METHOD = eval(value)
        if var == "stage2_absent_seconds":
            self.STAGE2_ABSENT_SECONDS = eval(value)
            if self.STAGE2_ABSENT_SECONDS > 0:
                self.stage2_action_enabled = True

    def usb_input_available(self, event):
        regex = re.compile(event)
        path = "/dev/input/by-id"
        exists = False

        for str in os.listdir(path):
            match = regex.search(str)
            if match:
                exists = True
                break
        return exists

    def load_monitors_config(self, monitor, items):
        if monitor == "ProcessMonitor" or monitor == "IOMonitor":
            self.MONITORS.append({"monitor":monitor, "name":items[0], "regex":eval(items[1]), "absent":self.ABSENT_SECONDS})
        if monitor == "InputMonitor" and (items[1] == "y" or items[1] == "yes"):
            if items[0] == "mouse" and self.usb_input_available("mouse"):
                self.MONITORS.append({"monitor":monitor, "name":items[0], "regex":"mice"})
            elif items[0] == "keyboard" and self.usb_input_available("kbd"):
                self.MONITORS.append({"monitor":monitor, "name":items[0], "regex":"kbd"})
            #else:
            #    self.MONITORS.append({"monitor":monitor, "name":items[0], "regex":items[1]})
        if monitor == "ConsoleMonitor" and (items[1] == "y" or items[1] == "yes"):
            self.MONITORS.append({"monitor":monitor, "name":items[0]})
        if monitor == "LoadMonitor":
            self.MONITORS.append({"monitor":monitor, "name":items[0], "threshold":items[1]})
        if monitor == "TCPMonitor":
            self.MONITORS.append({"monitor":monitor, "name":items[0], "port":items[1], "absent":self.ABSENT_SECONDS})
        if monitor == "UDPMonitor":
            # If ACTION_METHOD is 0 (PowerSave) and port is 7 or 9, do *NOT* create a monitor
            # This will cause that the WoL monitor to not be able to bind the port or viceversa.
            # TODO: Display a message that port is not being binded!!
            if self.ACTION_METHOD == 0 and (items[1] != 7 or items[1] != 9):
                self.MONITORS.append({"monitor":monitor, "name":items[0], "port":eval(items[1]), "absent":self.ABSENT_SECONDS})
        if monitor == "WoLMonitor":
            self.MONITORS.append({"monitor":monitor, "name":items[0], "port":eval(items[1]), "absent":self.ABSENT_SECONDS})
        if monitor == "DiskMonitor" and (items[1] == "y" or items[1] == "yes"):
            self.MONITORS.append({"monitor":monitor, "name":items[0], "absent":self.ABSENT_SECONDS})

    def get_monitors(self):
        monitor = []
        for config in self.MONITORS:
            if config["monitor"] == "ProcessMonitor":
                p = ProcessMonitor.ProcessMonitor(config)
            if config["monitor"] == "LoadMonitor":
                p = LoadMonitor.LoadMonitor(config)
            if config["monitor"] == "UDPMonitor":
                p = UDPMonitor.UDPMonitor(config)
            if config["monitor"] == "WoLMonitor":
                p = WoLMonitor.WoLMonitor(config)
            if config["monitor"] == "InputMonitor":
                p = InputMonitor.InputMonitor(config)
            if config["monitor"] == "ConsoleMonitor":
                p = ConsoleMonitor.ConsoleMonitor(config)
            if config["monitor"] == "IOMonitor":
                p = IOMonitor.IOMonitor(config)
            if config["monitor"] == "TCPMonitor":
                p = TCPMonitor.TCPMonitor(config)
            if config["monitor"] == "DiskMonitor":
                p = DiskMonitor.DiskMonitor(config)
            monitor.append(p)

        return monitor
